File: amazon/merge_two_linked_lists.py


def createLinkedList(listOfElements):
    headPointer = None
    pointer = None
    for item in listOfElements:
        if(headPointer == None):
            headPointer = LinkedList()
            pointer = headPointer
        else:
            pointer.nextNode = LinkedList()
            pointer = pointer.nextNode
        pointer.val = item
    return headPointer

        

class LinkedList:
    def __init__(self):
        self.val = 0
        self.nextNode = None

def mergeTwoLists(head1, head2):
    pointer1 = head1
    pointer2 = head2

    mergeList=None
    mergePointer=None

    count = 0
    while(count < 100):
        if pointer1 == None and pointer2 == None:
            break
        if pointer1 == None:
            if(mergeList == None):
                mergeList = LinkedList()
                mergeList = pointer2
            else:
                mergePointer.nextNode = pointer2
            break
        if pointer2 == None:
            if(mergeList == None):
                mergeList = LinkedList()
                mergeList = pointer1
            else:
                mergePointer.nextNode = pointer1
            break


        if pointer1.val > pointer2.val:
            if(mergeList == None):
                mergeList = pointer2
                mergePointer = mergeList
            else:
                mergePointer.nextNode = pointer2
                mergePointer = mergePointer.nextNode

            pointer2 = pointer2.nextNode
        else:
            if(mergeList == None):
                mergeList = pointer1
                mergePointer = mergeList
            else:
                mergePointer.nextNode = pointer1
                mergePointer = mergePointer.nextNode
            pointer1 = pointer1.nextNode
        count+=1
    return mergeList
            
            
            

def displayLinkedList(linkedList):
    while(linkedList != None):
        print(linkedList.val)
        linkedList = linkedList.nextNode


def main():
    list1 = [1,2,3,4,5,6,7,8]
    list2 = [3,44,55,66,88]
    list1 = createLinkedList(list1)
    list2 = createLinkedList(list2)

    mergedList = mergeTwoLists(list1, list2)
    
    displayLinkedList(mergedList)


if __name__ == "__main__": main()

File: amazon/merge_two_linked_lists.cpp
#include<iostream>

using namespace std;

struct LinkedListNode{
    int val;
    struct LinkedListNode * next;
};


typedef struct LinkedListNode node;


void display_list(node * pointer);

void createLinkedList(node * & head, int size){
    int value;
    node * pointer;
    for(int i = 0; i < size ; i++){
        cin>>value;
        if(head == NULL){
            head = new node();
            pointer = head;
            }
        else{
            pointer->next = new node();
            if(pointer->next == NULL){
                cout<<"pointer initialized points to garbage"<<endl;
            }
            pointer = pointer->next;
        }
        pointer->val = value;
        pointer->next = NULL;    
    }
}


node * mergeLinkedLists(node * pointer1, node * pointer2){
    node * mergePointer;
    node * mergedList = NULL;
    display_list(pointer1);
    display_list(pointer2);
    if(mergedList == NULL){
        cout<<"mergeList variable declare"<<endl;
    }
    while(1){
        if(pointer1 == NULL && pointer2 == NULL)
            break;
        else if(pointer1 == NULL){
            if(mergedList == NULL)
                mergedList = pointer2;
            else
                mergePointer->next = pointer2;
            break;
        }
        else if(pointer2 == NULL){
            if(mergedList == NULL)
                mergedList = pointer1;
            else
                mergePointer->next = pointer1;
            break;
        }
        else if(mergedList == NULL){
            if(pointer1->val < pointer2->val){
                mergedList= pointer1;
                mergePointer = mergedList;
                pointer1 = pointer1->next;
            }
            else{
                mergedList = pointer2;
                mergePointer = mergedList;
                pointer2 = pointer2->next;
            }
        }
        else{
            if(pointer2->val < pointer1->val){
                mergePointer->next = pointer2;
                pointer2 = pointer2->next;
                mergePointer = mergePointer->next;
            }
            else{
                mergePointer->next = pointer1;
                pointer1 = pointer1->next;
                mergePointer = mergePointer->next;

            }
        }
    }

    return mergedList;
}


void display_list(node * pointer){
    while(pointer != NULL){
        cout<<pointer->val<<endl;
        pointer = pointer->next;
    }
}

int main(){
    int size1, size2;
    node * head1;
    node * head2;
    cin>>size1>>size2;
    createLinkedList(head1, size1);
    createLinkedList(head2, size2);


    node * mergedList = mergeLinkedLists(head1, head2);

    display_list(mergedList);
    return 0;
}